src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+        /// </returns>
+        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, BoundExpression inPlaceTarget, out bool avoidInPlace, BoundExpression? start = null, BoundExpression? length = null)
         {
-            Debug.Assert(start is null == length is null);
+            Debug.Assert(start is null == length is null, "start and length must either both be null or both be non-null");
+            Debug.Assert(inPlaceTarget is null || TargetIsNotOnHeap(inPlaceTarget), "in-place construction target should not be on heap");
 
             ImmutableArray<byte> data = default;
             int elementCount = -1;
+            avoidInPlace = false;
+            SpecialType specialElementType = SpecialType.None;
 
             if (!_module.SupportsPrivateImplClass)
             {
                 return false;
             }
 
-            var ctor = (MethodSymbol?)Binder.GetWellKnownTypeMember(this._module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
-
-            if (ctor == null)
+            var rosPointerCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosPointerCtor is null)
             {
                 return false;
             }
-
-            Debug.Assert(!ctor.HasUnsupportedMetadata);
-
+            Debug.Assert(!rosPointerCtor.HasUnsupportedMetadata);
+
+            // The purpose of this optimization is to replace a BoundArrayCreation with better code generation.
+            // We're looking for an expression like:
+            //     new ReadOnlySpan<T>(new T[] { const, const, ... })
+            //     new ReadOnlySpan<T>(new T[] { const, const, ... }, 0, length)
+            //     (ReadOnlySpan<T>)new T[] { const, const, ... }
+            // etc., and wrappedExpression is that array creation.  For single byte primitives, we can replace that
+            // with the equivalent of:
+            //     new ReadOnlySpan<T>((void*)PrivateImplementationDetails.DataField, Length)
+            // on all target platforms.  For primitives larger than a single byte, if the target platform exposes
+            // the RuntimeHelpers.CreateSpan method, we can emit it instead as:
+            //     RuntimeHelpers.CreateSpan(PrivateImplementationDetails.DataFieldToken)
+            // and for platforms that lack CreateSpan, as a span that wraps a lazily-initialized array:
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= new T[] { ... })
+            // For non-constant data, unsupported primitive types, and other variations, this optimization will fail
+            // and the method will return false indicating that no code was emitted.
```

Comment:
Thanks a lot for the explanations (here and below in this method). It really helps to follow this method :-)